reps/2022-08-31-actor-affinity-apis.md

```diff
@@ -0,0 +1,405 @@
+#### GCS Scheduling Mode Implementation plan
+
+1. Actor adds the Labels property. Stored in the GcsActor structure
+2. Gcs Server add GcsLabelManager. Add labels->node information to GcsLabelManager after per actor completes scheduling.
+3. Added ActorAffinityPolicy, whose scheduling logic is to select the main scheduling node based on each matching expression and the labels->node information of GcsLabelManager
+
+The pseudo-code logic of the 4 matching operations is as follows:  
+Affinity Operator: IN
+```
+feasible_nodes = FindFeasibleNodes()
+for (expression : match_expressions) {
+    feasible_nodes = feasible_nodes ∩ GcsLabelManager->GetNodesByKeyAndValue(expression.key, expression.values)
+}
+```
+
+Affinity Operator: EXISTS
+```
+feasible_nodes = FindFeasibleNodes()
+for (expression : match_expressions) {
+    feasible_nodes = feasible_nodes ∩ GcsLabelManager->GetNodesByKey(expression.key)
+}
+```
+
+Affinity Operator: NOT_IN
+```
+feasible_nodes = FindFeasibleNodes()
+for (expression : match_expressions) {
+    feasible_nodes = feasible_nodes - GcsLabelManager->GetNodesByKeyAndValue(expression.key, expression. values)
+}
+```
+
+Affinity Operator: DOES_NOT_EXIST
+```
+feasible_nodes = FindFeasibleNodes()
+for (expression : match_expressions) {
+    feasible_nodes = feasible_nodes - GcsLabelManager->GetNodesByKey(expression.key)
+}
+```
+
+GcsLabelManger Desion  
+
+Add 4 public interfaces for GcsLabelManger :
+```c++
+// Add Actor's labels and node index relationship information.
+void AddActorLabels(const std::shared_ptr<GcsActor> &actor);
+
+// Remove Actor's labels and node index relationship information.
+void RemoveActorLabels(const std::shared_ptr<GcsActor> &actor);
+
+// Get the node where the actors those keys and values matched. 
+absl::flat_hash_set<NodeID> GetNodesByKeyAndValue(const std::string &ray_namespace, const std::string &key, const absl::flat_hash_set<std::string> &values) const;
+
+// Get the node where the actors which exist this key.
+absl::flat_hash_set<NodeID> GetNodesByKey(const std::string &ray_namespace, 
+                                          const std::string &key) const;
+```
+
+Main data structure :
+```
+Map<label_key, Map<lable_value, Set<node_id>>> label_to_nodes_
+Map<node_id, Set<GcsActor>> node_to_actors_
+```
+#### Raylet Scheduling Mode Implementation plan
+The implementation of Raylet scheduling mode is same as GCS scheduling above.  
+Mainly, one more Labels information needs to be synchronized to all Raylet nodes
+
+## Compatibility, Deprecation, and Migration Plan
+
+## Test Plan and Acceptance Criteria
+All APIs will be fully unit tested. All specifications in this documentation will be thoroughly tested at the unit-test level. The end-to-end flow will be tested within CI tests. Before the beta release, we will add large-scale testing to precisely understand scalability limitations and performance degradation in large clusters. 
+
+## (Optional) Follow-on Work
+
```

Comment:
We need to discuss more about the failure modes.

For example if actor B collocates with actor A:

What if actor A fails/out of scope, what should happen to actor B?
What if actor A fails and restarts on a different node, what should happen to actor B?

Reply:
OK, I will add to ”Failures and Special Scenarios“ chapter.  

If actor B collocates with actor A and The actor A & B both complete schedule to Node 1.
1、If actor A failed and restarted to Node 2, the actor B will not be affected, it still working in Node 1.  
2、If actor A failed and restarted to Node 2, Then the actor B also failed, It will be scheduled to Node 2 and collocates with actor A.